text/0000-node-api-v2.md

```diff
@@ -0,0 +1,994 @@
+
+- Title: Node API V2
+- Authors: [Quentin Le Sceller](mailto:q.lesceller@gmail.com)
+- Start date: September 30th, 2019
+- RFC PR: Edit if merged: [mimblewimble/grin-rfcs#0000](https://github.com/mimblewimble/grin-rfcs/pull/0000) 
+- Tracking issue: [Edit if merged with link to tracking github issue]
+
+---
+
+# Table of Contents
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+- [Community-level explanation](#community-level-explanation)
+- [Reference-level explanation](#reference-level-explanation)
+  * [Current Endpoints with the v1 API](#current-endpoints-with-the-v1-api)
+  * [Proposed Endpoints](#proposed-endpoints)
+    + [Miscellaneous endpoints](#miscellaneous-endpoints)
+      - [get_status](#get_status)
+      - [get_version](#get_version)
+    + [Chain endpoints](#chain-endpoints)
+      - [get_header](#get_header)
+      - [get_block](#get_block)
+      - [get_tip](#get_tip)
+      - [get_kernel](#get_kernel)
+      - [get_outputs](#get_outputs)
+      - [get_unspent_outputs](#get_unspent_outputs)
+      - [validate_chain](#validate_chain)
+      - [compact_chain](#compact_chain)
+    + [Peer endpoints](#peer-endpoints)
+      - [get_peers](#get_peers)
+      - [get_connected_peers](#get_connected_peers)
+      - [ban_peer](#ban_peer)
+      - [unban_peer](#unban_peer)
+    + [Pool endpoints](#pool-endpoints)
+      - [get_pool_size](#get_pool_size)
+      - [get_stempool_size](#get_stempool_size)
+      - [get_unconfirmed_transactions](#get_unconfirmed_transactions)
+      - [push_transaction](#push_transaction)
+  * [Authentication](#authentication)
+    + [Wallet support](#wallet-support)
+    + [Legacy support](#legacy-support)
+    + [API only](#api-only)
+- [Drawbacks](#drawbacks)
+- [Prior art](#prior-art)
+- [Future possibilities](#future-possibilities)
+- [References](#references)
+
+# Summary
+
+[summary]: #summary
+
+Create a v2 JSON-RPC API for the Node API.
+
+# Motivation
+
+[motivation]: #motivation
+
+The current Node API (referred here as v1 API) is a REST API. This API while simple is:
+
+- Documented manually (current documentation is available [here](https://github.com/mimblewimble/grin/blob/master/doc/api/node_api.md).
+- Contains call with heterogenous args such as `?byid=xxx` and `commitment/xxx` which can be confusing and lack some uniformity.
+- Uses REST which is bound to HTTP while v2 wallet API uses JSON-RPC.
+
+The goal of this RFC is to provide a new API with:
+- Cleaner methods and errors.
+- Generated documentation directly on docs.rs.
+- Automatic testing with doc tests.
+- Stronger basis for future improvements.
+
+# Community-level explanation
+
+[community-level-explanation]: #community-level-explanation
+
+This new API will be particularly useful for wallet developer and should ultimately simplify their work on Grin. Moreover, this RFC does not introduce any breaking changes as the v1 REST API will still be around until completely deprecated.
```

Comment:
Is there a timeframe for removing the old API? Can we take advantage of one of the next hardforks? In a world without hardforks, it would probably be useful to provide some API guarantees to clients, eg. in X.Y we guarantee no backward compatible changes until X.Z (where Z=Y+2, iow, the current APIs in use are guaranteed to not break for the next two releases).

Reply:
That'd be the idea here. Comments from everyone are more than welcome.

Reply:
It might make sense to include a timeline for deprecation of v1 as part of the RFC.